rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py

```diff
@@ -0,0 +1,267 @@
+import logging
+import os
+import typing
+import warnings
+from typing import Any, Dict, List, Optional, Text, Tuple, Type
+
+import numpy as np
+import scipy.sparse
+
+import rasa.shared.utils.io
+import rasa.utils.io as io_utils
+from rasa.shared.constants import DOCS_URL_TRAINING_DATA_NLU
+from rasa.nlu.classifiers import LABEL_RANKING_LENGTH
+from rasa.nlu.featurizers.featurizer import SparseFeaturizer
+from rasa.nlu.components import Component
+from rasa.nlu.classifiers.classifier import IntentClassifier
+from rasa.nlu.config import RasaNLUModelConfig
+from rasa.shared.nlu.constants import TEXT
+from rasa.nlu.model import Metadata
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.shared.nlu.training_data.message import Message
+
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    import sklearn
+
+
+class SparseSklearnIntentClassifier(IntentClassifier):
```

Comment:
Technically we may be implementing more sparse classifiers from sklearn. Maybe rename to `SparseNaiveBayesClassifier`?

Reply:
Yeah I can change the name. One thing I'm hazy on though, I think most scikit learn models can handle sparse data or dense data. Why not make a NaiveBayesClassifier, and give it an option flag to select whether to use sparse or dense features from the rasa feature extraction pipeline?

Reply:
Scikit-learn offers [many different kinds of NaiveBayes models](https://scikit-learn.org/stable/modules/naive_bayes.html) actually. I even maintain an open-source project that adds [yet another kind](https://scikit-lego.readthedocs.io/en/latest/naive-bayes.html) too. Scikit-learn implicitly translates sparse data whenever it is combined with dense data into dense data. It's a fair way of dealing with it, but it's not very performant. 

As far as Rasa users are concerned though, I think it's fair to assume that DIET will always be more performant in terms of accuracy. The main use-case for a sparse model here is to have a simple benchmark that is *much* faster to train. I'm currently trying to implement a form and I need to wait for 10-20s every time I iterate because DIET needs to train. It'd be nice if this was reduced.